src/eval.ml

```diff
@@ -187,34 +196,47 @@ and tree_walk : dtree -> stack -> (term * stack) option = fun tree stk ->
                 env.(slot) <- TE_Some(b)
               else
                 let b = lift vars.(pos) in
-                let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+                let xs = Array.map (fun e -> IntMap.find e id_vars) xs in
                 env.(slot) <- TE_Some(Bindlib.unbox (Bindlib.bind_mvar xs b))
         in
         List.iter fn env_builder;
         (* Actually perform the action. *)
         Some(Bindlib.msubst act env, stk)
-    | Cond({ ok ; cond ; fail })                          ->
+    | Cond({ok; cond; fail})                              ->
         let next =
           match cond with
           | CondNL(i, j) ->
               if eq_modulo vars.(i) vars.(j) then ok else fail
-          | CondFV(xs,i) ->
-              let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+          | CondFV(i,xs) ->
+              let allowed =
+                (* Variables that are allowed in the term. *)
+                let fn id =
+                  try IntMap.find id id_vars with Not_found -> assert false
+                in
+                Array.map fn xs
+              in
+              (* Ensure there are no variables from [forbidden] in [b]. *)
+              let no_forbidden b =
+                let forbidden =
+                  (* Term variables forbidden in the term. *)
+                  IntMap.filter (fun id _ -> not (Array.mem id xs)) id_vars
+                in
+                IntMap.for_all (fun _ x -> not (Bindlib.occur x b)) forbidden
```

Comment:
Why do you prefer `IntMap.for_all (fun _ x -> not (f x)) m` with O(m) calls to `not` instead of `not (IntMap.exists (fun _ -> f) m)` with 1 call to `not`?

Reply:
Ah it just seemed more natural to me, but this is very personal. I'll change this.